includes/acl/clip_reader/clip_reader.h

```diff
@@ -0,0 +1,450 @@
+				m_allocator.deallocate(m_skeleton);
+				m_skeleton = nullptr;
+			}
+
+			return false;
+		}
+
+		AnimationClip* get_clip()
+		{
+			return m_clip;
+		}
+
+		RigidSkeleton* get_skeleton()
+		{
+			return m_skeleton;
+		}
+
+		ClipReaderError get_error()
+		{
+			return m_error;
+		}
+
+	private:
+		// Temporary storage for input that must be translated to compressor data structures.
+		struct Header
+		{
+			const char* clip_name;
+			int clip_name_length;
+			int num_samples;
+			int sample_rate;
+			double error_threshold;
+			const char* reference_frame;
+			int reference_frame_length;
+		};
+
+		struct Bone
+		{
+			int index;
+			const char* name;
+			int name_length;
+			const char* parent;
+			int parent_length;
+			double vertex_distance;
+			double bind_rotation[4];
+			double bind_translation[3];
+			double bind_scale[3];
+		};
+
+		Allocator& m_allocator;
+		bool m_read_already{};
+		SJSONParser m_parser;
+		ClipReaderError m_error;
+
+		RigidSkeleton* m_skeleton{};
+		AnimationClip* m_clip{};
+
+		double m_version{};
+		Header m_header{};
+
+		int m_num_bones{};
+		char* m_bone_names{};
+
+		void set_error(int reason)
+		{
+			m_parser.get_position(m_error.line, m_error.column);
+			m_error.error = reason;
+		}
+
+		bool read_version()
+		{
+			if (!m_parser.read("version", m_version))
+			{
+				m_error = m_parser.get_error();
+				return false;
+			}
+
+			if (m_version != 1.0)
+			{
+				set_error(ClipReaderError::UnsupportedVersion);
+				return false;
+			}
+
+			return true;
+		}
+
+		bool read_clip_header()
+		{
+			if (!m_parser.object_begins("clip"))
+			{
+				goto error;
+			}
+
+			if (!m_parser.read("name", m_header.clip_name, m_header.clip_name_length) ||
+				!m_parser.read("num_samples", m_header.num_samples) ||
+				!m_parser.read("sample_rate", m_header.sample_rate) ||
+				!m_parser.read("error_threshold", m_header.error_threshold))
+			{
+				goto error;
+			}
+
+			m_parser.try_read("reference_frame", m_header.reference_frame, m_header.reference_frame_length);
+
+			if (!m_parser.object_ends())
+			{
+				goto error;
+			}
+
+			return true;
+
+		error:
+			m_error = m_parser.get_error();
+			return false;
+		}
+
+		bool read_and_allocate_bones()
+		{
+			int name_buffer_length = 0;
+			RigidBone* bones = nullptr;
+
+			SJSONParser::State before_bones = m_parser.save_state();
+
+			if (!for_each_bone([this, &name_buffer_length](Bone input)
+			{
+				++m_num_bones;
+				name_buffer_length += input.name_length + 1;
+				return true;
+			}))
+			{
+				goto error;
+			}
+
+			bones = allocate_type_array<RigidBone>(m_allocator, m_num_bones);
+
+			m_bone_names = allocate_type_array<char>(m_allocator, name_buffer_length);
+			char* current_name = m_bone_names;
+
+			m_parser.restore_state(before_bones);
+
+			if (!for_each_bone([bones, &current_name](Bone src)
+			{
+				RigidBone* dst = bones + src.index;
+
+				std::memcpy(current_name, src.name, src.name_length);
+				current_name[src.name_length] = '\0';
```

Comment:
but the names live in the sjson memory buffer as well, we could easily use const char* to point into it and return the length when we parse it (since it isn't null terminated) a bit of a pain, perhaps something like a SJSONStringRef